base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+            )
+            .concat(
+                (&self.state_data.sender_excess.unwrap() + &self.state_data.receiver_output_public_key.unwrap())
+                    .as_bytes(),
+            )
+            .concat(&self.state_data.fee.unwrap().to_le_bytes())
+            .concat(&self.state_data.lock_height.unwrap().to_le_bytes());
+
+        self.state_data.receiver_partial_signature = Some(Signature::sign(
+            self.state_data.receiver_output_blinding_factor.unwrap(),
+            self.state_data.receiver_private_nonce.unwrap(),
+            challenge.clone(),
+        )?);
+
+        Ok(TransactionProtocolState::SenderPartialSignature(SenderPartialSignature::new(self.state_data)))
+    }
+}
+
+/// In this state the Sender can now construct their partial signature and the final aggregated signature.
+/// Also in this state the Receiver can construct the message containing their public data to be sent back to the Sender
+#[derive(Clone, Debug)]
+struct SenderPartialSignature {
+    state_data: TransactionProtocolStateData,
+}
+
+impl SenderPartialSignature {
+    fn new(previous_state_data: TransactionProtocolStateData) -> SenderPartialSignature {
+        SenderPartialSignature { state_data: previous_state_data }
+    }
+
+    fn construct_receiver_public_data(&self) -> Result<ReceiverPublicData, TransactionProtocolError> {
+        // Validate the current state to check we have the data we need
+        if self.state_data.receiver_output_public_key.is_none() ||
+            self.state_data.receiver_public_nonce.is_none() ||
+            self.state_data.receiver_partial_signature.is_none() ||
+            self.state_data.receiver_output.is_none()
+        {
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        Ok(ReceiverPublicData {
+            receiver_output: self.state_data.receiver_output.unwrap(),
+            receiver_output_public_key: self.state_data.receiver_output_public_key.unwrap(),
+            receiver_public_nonce: self.state_data.receiver_public_nonce.unwrap(),
+            receiver_partial_signature: self.state_data.receiver_partial_signature.unwrap(),
+        })
+    }
+
+    fn finalize_signature(mut self) -> Result<TransactionProtocolState, TransactionProtocolError> {
+        // Validate that all the required state is present.
+        if self.state_data.sender_public_nonce.is_none() ||
+            self.state_data.sender_private_nonce.is_none() ||
+            self.state_data.sender_excess.is_none() ||
+            self.state_data.receiver_output_public_key.is_none() ||
+            self.state_data.fee.is_none() ||
+            self.state_data.lock_height.is_none() ||
+            self.state_data.receiver_public_nonce.is_none() ||
+            self.state_data.receiver_partial_signature.is_none()
+        {
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        let challenge = Challenge::<Blake256>::new()
```

Comment:
Could this be DRYed up? It looks the same as the code in the previous function

Reply:
Made a function for it